packages/trpc/src/procedures/auth.ts

```diff
@@ -12,6 +13,126 @@ import {
 const getNewSessionExpirationDate = () =>
   new Date(Date.now() + 60_000 * 60 * 24 * 30);
 
+const getNewEmailVerificationCodeExpirationDate = () =>
+  new Date(Date.now() + 60_000 * 15);
+
+export const sendEmailVerificationProcedure = notAuthenticatedProcedureBuilder
+  .input(
+    z.object({
+      email: z.string().email(),
+    }),
+  )
+  .mutation(async ({ ctx, input }) => {
+    // Check if there is already an email verification code for the given email
+    const existingEmailVerificationCode =
+      await ctx.prisma.emailVerificationCode.findUnique({
+        where: {
+          email: input.email,
+        },
+      });
+    // If email already verified, throw error
+    if (existingEmailVerificationCode?.emailConfirmed) {
+      throw new TRPCError({
+        code: "BAD_REQUEST",
+        message: "Email already verified",
+      });
+    }
+    // If email not verified, delete current email verification code to create a new one
+    if (existingEmailVerificationCode !== null) {
+      await ctx.prisma.emailVerificationCode.delete({
+        where: {
+          email: input.email,
+        },
+      });
+    }
+
+    // Generate 6 digit code for email verification
+    let emailCode = "";
+    for (let i = 0; i < 6; i++) {
+      emailCode += Math.floor(Math.random() * 10);
```

Comment:
Might be worth splitting the code generation here out to a standalone function in the email service. This way we can mock it in testing (and thus remove randomness where needed)